linera-service/src/exporter/runloops/mod.rs

```diff
@@ -0,0 +1,190 @@
+            let moved_storage = storage.clone()?;
+            let destinations = destination_config
+                .destinations
+                .iter()
+                .enumerate()
+                .filter_map(|(i, dest)| {
+                    if i % number_of_threads == n {
+                        Some((i as u16, dest.clone()))
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            let handle = thread::spawn(move || {
+                start_exporter_tasks(
+                    moved_signal,
+                    options,
+                    work_queue_size,
+                    moved_storage,
+                    destinations,
+                )
+            });
+
+            pool.push(handle);
+        }
+
+        pool
+    };
+
+    Ok(threadpool)
+}
+
+#[tokio::main(flavor = "current_thread")]
+async fn start_exporter_tasks<S, F>(
+    shutdown_signal: F,
+    options: NodeOptions,
+    work_queue_size: usize,
+    mut storage: ExporterStorage<S>,
+    destinations: Vec<(DestinationId, Destination)>,
+) where
+    S: Storage + Clone + Send + Sync + 'static,
+    F: IntoFuture<Output = ()> + Clone + Send + Sync + 'static,
+    <F as IntoFuture>::IntoFuture: Future<Output = ()> + Send + Sync + 'static,
+{
+    let node_provider = GrpcNodeProvider::new(options);
+    let arced_node_provider = Arc::new(node_provider);
+    let mut set = JoinSet::new();
+    for (id, destination) in destinations {
+        let exporter_task = Exporter::new(
+            arced_node_provider.clone(),
+            id,
+            storage.clone().unwrap(),
+            destination,
+            work_queue_size,
+        );
+        let _handle = set.spawn(exporter_task.run_with_shutdown(shutdown_signal.clone()));
+    }
+
+    while let Some(res) = set.join_next().await {
+        res.unwrap().unwrap();
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+#[tokio::main(flavor = "current_thread")]
+async fn start_block_processor<S, F>(
+    storage: S,
+    shutdown_signal: F,
+    limits: LimitsConfig,
+    options: NodeOptions,
+    block_exporter_id: u32,
+    clients_per_thread: usize,
+    queue_rear: UnboundedSender<BlockId>,
+    destination_config: DestinationConfig,
+    queue_front: UnboundedReceiver<BlockId>,
+) -> Result<(), ExporterError>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+    F: IntoFuture<Output = ()> + Clone + Send + Sync + 'static,
+    <F as IntoFuture>::IntoFuture: Future<Output = ()> + Send + Sync + 'static,
+{
+    let (block_processor_storage, exporter_storage) = BlockProcessorStorage::load(
+        storage.clone(),
+        block_exporter_id,
+        destination_config.destinations.len() as u16,
+        limits,
+    )
+    .await?;
+
+    let mut block_processor = BlockProcessor::new(block_processor_storage, queue_rear, queue_front);
+
+    let _pool = start_exporters(
```

Comment:
You collect the thread handles inside `start_exporters` nicely now but drop it here - why not return to the caller?

Reply:
NVM, this method doesn't return unless `block_processor.run_with_shutdown` finishes anyway.